### PR TITLE
[aotd] Fix freezing API for subclasses

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136265

Original issue: 
https://github.com/pytorch/ao/issues/890

The problem:

TracingContext.flat_params contain original params, with not desugared Subclasses.
While inductor.freezing API works on aot graphs, which already desugared Subclasses.

flat_params are used only for this logic and storing in them desguared subclasses fixes the issue.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_inductor_freezing_with_subclasses
```
Torch AO original failure:
```
python test/integration/test_integration.py -k test_int8_weight_only_quant_with_freeze
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -977,7 +977,9 @@ def aot_module_simplified(
     full_args.extend(params_flat)
 
     if tracing_context := torch._guards.TracingContext.try_get():
-        tracing_context.params_flat = params_flat
+        tracing_context.params_flat = unwrap_tensor_subclasses(
+            params_flat, is_joint_structure=False
+        )
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:982]: This does seem directionally right. also cc @mlazos

**Review by bdhirsh — APPROVED** [review]
sgtm!

**Comment by Valentine233** [comment]
When applying to model zoo, we also find the same error of `IndexError: list index out of range` as CI.

Is it reasonable to do the following modification?
From `max_offset_idx = max(0, len(params_flat_unwrap) - 1)` to `max_offset_idx = max(0, len(unwrapped_args_offsets) - 1)`.

**Pull request closed without merging**